### PR TITLE
adapter: Add resource limit for compute credits

### LINKED ISSUE #18830: Add `max_credits_per_hour` system limit
### Feature request

Add a new `max_credits_per_hour` system limit, which restricts the maximum hourly credit burn of an environment.

Detailed proposal: https://www.notion.so/materialize/Self-guided-trial-viability-47c4b14f19b44c58a630d360a7e615a8?pvs=4#1c493d1fa3a94483a5da54f61f1e4f3c

Timeline: to hit next week's go live date, we need this limit to land in this week's release (i.e., by Friday 4/21 at 2pm ET).

cc @chaas @hlburak

### PR DESCRIPTION
**Pull request by jkosh44**
Adds a resource limit for compute credits pre hour. Additionally, update the resource limit error messages.

Resolves #18830

### Motivation
This PR adds a known-desirable feature.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label (https://github.com/MaterializeInc/cloud/issues/6216).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - This release adds a max compute credits per hour resource limit

### CODE REVIEW

**Comment by jkosh44** [comment]
@philip-stoev Feel free to delegate to someone else on the QA team. Also is there a better way to get the tests passing other than sprinkling a bunch of `ALTER SYSTEM SET` everywhere?

File: src/controller/src/clusters.rs
Comment (umanwizard) [inline src/controller/src/clusters.rs:83]: It feels odd that this is a float (instead of e.g. an int representing 1000ths of a credit). I've traditionally been taught that anything accounting-related being a float is a code smell.

**Comment by umanwizard** [comment]
Don't we need a cloud PR to add the new values to the size mappings?

**Comment by umanwizard** [comment]
We should have an issue with the release-blocker tag for this to be added to cloud (per the PR checklist)

File: src/controller/src/clusters.rs
Comment (jkosh44) [reply src/controller/src/clusters.rs:83]: What about `Numeric` (i.e. `Decimal<13>`), does that have the same code smell?

Comment (umanwizard) [reply src/controller/src/clusters.rs:83]: AFAIK no

Comment (jkosh44) [reply src/controller/src/clusters.rs]: I switched to `Numeric` then. I think it's a bit more straight forward than converting back and forth between 1000ths of a credit.

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:2297]:
```suggestion
    const TYPE_NAME: &'static str = "numeric";
```

File: src/adapter/src/catalog/config.rs
Comment (benesch) [inline src/adapter/src/catalog/config.rs:151]:
```suggestion
                credits_per_hour: (2 * 4).into(),
```

File: src/controller/src/clusters.rs
Comment (benesch) [inline src/controller/src/clusters.rs:84]:
```suggestion
    pub credits_per_hour: Numeric,
```

File: src/sql/src/session/vars.rs
Comment (benesch) [inline src/sql/src/session/vars.rs:380]:
```suggestion
    name: UncasedStr::new("max_credits_per_hour"),
```

**Review by benesch — APPROVED** [review]
LGTM modulo bikeshed about `max_credits_per_hour`. 

Maybe 64 was too low of a default. Should we set it to 1024 by default so you don't have to constantly adjust tests to bump the limit?

**Comment by benesch** [comment]
Thanks again for turning this around so quickly!

**Review by mjibson — APPROVED** [review]

File: src/sql/src/session/vars.rs
Comment (benesch) [reply src/sql/src/session/vars.rs:382]: Ah, no, "credits" are for replicas only! Storage is called "storage" and network egress is called "network egress"!

image

File: src/adapter/src/catalog/config.rs
Comment (philip-stoev) [inline src/adapter/src/catalog/config.rs:113]: I think this should count as 1 credit, given that scale = 1 .

File: src/adapter/src/catalog/config.rs
Comment (philip-stoev) [inline src/adapter/src/catalog/config.rs:139]: This should be `scale` only I think.

File: src/adapter/src/catalog/config.rs
Comment (philip-stoev) [inline src/adapter/src/catalog/config.rs:151]: credits per hour = 2 here I think.

File: test/cluster/resources/resource-limits.td
Comment (philip-stoev) [inline test/cluster/resources/resource-limits.td:370]: Use size 2-1 here to get a 2-credit replica

File: test/cluster/resources/resource-limits.td
Comment (philip-stoev) [inline test/cluster/resources/resource-limits.td:374]: This also needs to be tested that it errors out:

```
 CREATE SOURCE lg2 FROM LOAD GENERATOR COUNTER WITH (SIZE = '8-8');
```

**Comment by philip-stoev** [comment]
Item no 1 : Do we need a `SHOW` command that will show the current credit usage? Otherwise people will be forced to 
issue a `CREATE` DDL in order to obtain the `current: XXXX` value from the error message.

Unlike the other limits (max_tables, max_sources), the current consumption can not be easily calculated using a SQL SELECT query.

**Comment by jkosh44** [comment]
> Item no 1 : Do we need a `SHOW` command that will show the current credit usage? Otherwise people will be forced to issue a `CREATE` DDL in order to obtain the `current: XXXX` value from the error message.
> 
> Unlike the other limits (max_tables, max_sources), the current consumption can not be easily calculated using a SQL SELECT query.

How about adding this as a column to `SHOW CLUSTER REPLICAS`? @chaas do you have thoughts on that?

File: src/adapter/src/catalog/config.rs
Comment (jkosh44) [reply src/adapter/src/catalog/config.rs:139]: I know nothing about how credits per hour are calculated, what's the rationale of just using `scale` instead of `scale * worker`?

Comment (philip-stoev) [reply src/adapter/src/catalog/config.rs:139]: This entire code here is only used for Mz instances that are used for testing purposes. I would say that all workers that run on the same machine should count as a single credit. This makes reasoning about tests easier I think. But I do not really have a very strong opinion on this.

**Comment by philip-stoev** [comment]
Item No 2. Can you please run Nightly against this -- I suspect that the CI jobs that use a non-default SIZE for the default cluster or a non-default number of replicas will fail. You may need to add a $skip-if directive that prevents this part of the test from running in those jobs:

```
$ skip-if
SELECT '${arg.default-replica-size}' != '4-4' OR ${arg.replicas} > 1;
```

File: src/adapter/src/catalog/config.rs
Comment (jkosh44) [reply src/adapter/src/catalog/config.rs]: If it makes testing easier, then I think it's worth it. I just pushed the changes.

**Comment by jkosh44** [comment]
> Item No 2. Can you please run Nightly against this -- I suspect that the CI jobs that use a non-default SIZE for the default cluster or a non-default number of replicas will fail. You may need to add a $skip-if directive that prevents this part of the test from running in those jobs:
> 
> ```
> $ skip-if
> SELECT '${arg.default-replica-size}' != '4-4' OR ${arg.replicas} > 1;
> ```

https://buildkite.com/materialize/nightlies/builds/2237

**Comment by philip-stoev** [comment]
Item No 3. The settings has a "per hour" in its name, but is there anything timing dependent here? From what I understand, it limits the current usage at any given time, that is, it is not possible to have 60 concurrent clusters running for 1 minute and consume 1 credit :-) Some customer confusion may arise as a result.

**Review by philip-stoev — APPROVED** [review]
Thank you for the tests! I played with the feature manually and the accounting was correct at all times.

**Comment by chaas** [comment]
> Item no 1 : Do we need a `SHOW` command that will show the current credit usage? Otherwise people will be forced to issue a `CREATE` DDL in order to obtain the `current: XXXX` value from the error message.
> 
> Unlike the other limits (max_tables, max_sources), the current consumption can not be easily calculated using a SQL SELECT query.

We can suggest users run this query?
```
SELECT SUM(s.credits_per_hour) total_credits_per_hour from mz_clusters c
JOIN mz_cluster_replicas r ON c.id = r.cluster_id
JOIN mz_internal.mz_cluster_replica_sizes s ON r.size=s.size;
```

I'm a little hesitant to add information into our SHOW commands that we're not confident to stabilize. We do already have the `ready` column though, so there is some precedent to adding information from `mz_cluster_replica_sizes`, but I'm inclined to see if the above query is sufficient for now.

**Comment by jkosh44** [comment]
I'm going to merge this once the cloud PR is merged.

**Comment by benesch** [comment]
> Item No 3. The settings has a "per hour" in its name, but is there anything timing dependent here? From what I understand, it limits the current usage at any given time, that is, it is not possible to have 60 concurrent clusters running for 1 minute and consume 1 credit :-) Some customer confusion may arise as a result.

> This is hard to do well because it ties into how we charge, but I agree that `per_hour` in the name makes this even harder. If this isn't a limit per hour, but rather a limit at any given point in time, like `max_databases` or `max_clusters`, I would remove `per_hour` from the name.

The problem is that the _unit_ is not "credits" but "credits per hour." I agree this is confusing!

Here's how I think about it: this limit is equivalent to a driving speed limit. We call that a "speed limit" or a "maximum speed", but the limit is expressed in miles per hour. You might even say "max mph." But you wouldn't want to say "max miles", as that's would express a constraint on the maximum distance traveled, not the maximum rate at which distance is traveled (speed).

Would `max_credit_rate` be less confusing? (Or ... `max_credit_speed`?)

**Comment by benesch** [comment]
The problem with `max_credit_rate` is that that "credit rate" sounds like "the price I pay for credits" not "the rate at which I am depleting credits." Sigh. Naming is hard.

**Comment by sploiselle** [comment]
`max_credit_burn`? Or too negative?

**Comment by benesch** [comment]
Filing as a release blocker because this needs to make the release tomorrow.

**Comment by jkosh44** [comment]
Here are the current proposals (plus two that I'm adding):

- `max_credits_per_hour`
- `max_credit_rate`
- `max_credit_burn`
- `max_credit_limit`
- `credit_limit`

@benesch Do you want to just have the final say and pick one? We can always change it later since users never actually type in the field name.

Also,

> Also, is this really per region and not per account?

Can you confirm this? I'm pretty sure the answer is yes.

**Comment by benesch** [comment]
> > Also, is this really per region and not per account?
> 
> Can you confirm this? I'm pretty sure the answer is yes.

Yes, confirmed.

**Comment by benesch** [comment]
> @benesch Do you want to just have the final say and pick one?

Well, I like what we currently have (`max_credits_per_hour`)—after all I proposed it! But @jseldess is the one who has to explain it. 

One more proposal to add: `max_credit_consumption_rate`. 

I say we give it until 11am tomorrow for @jseldess and/or @frankmcsherry to express an opinion. If we don't hear from them, then let's go with what you have.

**Comment by jseldess** [comment]
> The problem is that the unit is not "credits" but "credits per hour." I agree this is confusing!

@benesch, I definitely don't want to block, but I I'm still having trouble understanding this. The credits per hour unit is for billing, correct? So if in a given hour I have 2 medium replicas (4 credits per hour each), for the full hour or any portion of it, we bill for 8 credits? 

Assuming that's how it works, and it's probably not, this internal limit isn't using the unit the same way. Here, it's not for billing per hour but for limiting the number of resources we allow at any given point in time. Am I way off?

**Comment by benesch** [comment]
> @benesch, I definitely don't want to block, but I I'm still having trouble understanding this. The credits per hour unit is for billing, correct? So if in a given hour I have 2 medium replicas (4 credits per hour each), for the full hour or any portion of it, we bill for 8 credits?

So, we bill by the second. Say you have those 2 medium replicas for 3 minutes each. We'll bill you for 2 replicas * 4 credits/hour * 3/60 minutes/hour = 0.4 credits.

The limit restricts the maximum instantaneous credit consumption rate. So if `max_credits_per_hour` were set to 6, you'd be allowed to have _one_ medium replica (instantaneous credit consumption rate of 4 credits per hour), but not _two_, as that would result in an instantaneous credit consumption rate of 8 credits per hour, exceeding the maximum consumption rate of 6. But you could have one medium replica and one small replica, as that would be a instantaneous credit consumption rate of exactly 6 credits per hour.

**Comment by benesch** [comment]
I am increasingly thinking that `max_credit_consumption_rate` is more clear than `max_credits_per_hour`.

**Comment by jkosh44** [comment]
> I say we give it until 11am tomorrow for @jseldess and/or @frankmcsherry to express an opinion. If we don't hear from them, then let's go with what you have. Does that timing work for you, @jkosh44?

Yep, that gives me 3 hours before the 2pm deadline which should be plenty.

**Comment by philip-stoev** [comment]
> We can suggest users run this query?

> SELECT SUM(s.credits_per_hour) total_credits_per_hour from mz_clusters c
> JOIN mz_cluster_replicas r ON c.id = r.cluster_id
> JOIN mz_internal.mz_cluster_replica_sizes s ON r.size=s.size;

oh, I did not realize this column exists -- it does not seem to be exercised by the test suite.

Apart from that, the query is reasonable, (**except that it also counts the internal clusters, which the limit enforcement does not**) so I think no further changes to `SHOW` or any of the other tables would be warranted.

**Comment by philip-stoev** [comment]
I suspected this PR will have a merge skew conflict with a test-fixing PR that operated in the same area, so I kicked the merge skew checker and indeed it is now red :

https://buildkite.com/materialize/tests/builds/54145#0187a3fd-4a99-42bc-b3a2-7e8550a99006

**Comment by jseldess** [comment]
Sorry to be so down to the wire here. I think it's hard to find a name that fully captures the meaning here, but I like `max_credit_consumption_rate` better than `max_credits_per_hour`. If we go with that, could we change the description to the following?

> The maximum rate of credit consumption in a region. Credits are consumed based on the size of cluster replicas in use.

Lots that still needs to be unpacked, but probably not (just) in this variable description.

**Comment by jkosh44** [comment]
Ok, it's 11AM. I went with the name `max_credit_consumption_rate` and description "The maximum rate of credit consumption in a region. Credits are consumed based on the size of cluster replicas in use.". I'm setting this to auto-merge, so speak up and disable auto-merge if you have something to say.

**Pull request merged**
